### PR TITLE
feat: Add snowflake grant ownership resource

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
TODO

## Test Plan

* [ ] acceptance tests

* [ ] …

## References

*

### CODE REVIEW

File: pkg/resources/grant_ownership.go [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+			},
+		},
+	}
+}
+
+func GrantOwnership() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: CreateGrantOwnership,
+		DeleteContext: DeleteGrantOwnership,
+		ReadContext:   ReadGrantOwnership,
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:212]: let's leave a comment that update is not needed because everything is forceNew. Also, should we have (is there already in the SDK) a no-op function we can just reference in such cases?

File: pkg/resources/grant_ownership.go [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+	},
+}
+
+func grantOwnershipBulkOperationSchema(branchName string) map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"object_type_plural": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			Description:  "Specifies the type of object in plural form on which you are transferring ownership. Available values are: AGGREGATION POLICIES | ALERTS | AUTHENTICATION POLICIES | COMPUTE POOLS | DATABASES | DATABASE ROLES | DYNAMIC TABLES | EVENT TABLES | EXTERNAL TABLES | EXTERNAL VOLUMES | FAILOVER GROUPS | FILE FORMATS | FUNCTIONS | HYBRID TABLES | ICEBERG TABLES | IMAGE REPOSITORIES | INTEGRATIONS | MATERIALIZED VIEWS | NETWORK POLICIES | NETWORK RULES | PACKAGES POLICIES | PIPES | PROCEDURES | MASKING POLICIES | PASSWORD POLICIES | PROJECTION POLICIES | REPLICATION GROUPS | ROLES | ROW ACCESS POLICIES | SCHEMAS | SESSION POLICIES | SECRETS | SEQUENCES | STAGES | STREAMS | TABLES | TAGS | TASKS | USERS | VIEWS | WAREHOUSES",
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:180]: same as above

File: pkg/resources/grant_ownership.go [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+			}
+		}
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func ReadGrantOwnership(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:379]: so are there cases where we should setId to `""` in read like in the normal resources?

File: pkg/resources/grant_ownership.go [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+	}
+
+	return ownershipGrantTo
+}
+
+func getOwnershipGrantOpts(id *GrantOwnershipId) *sdk.GrantOwnershipOptions {
+	if id != nil && id.OutboundPrivilegesBehavior != nil {
+		outboundPrivileges := id.OutboundPrivilegesBehavior.ToOwnershipCurrentGrantsOutboundPrivileges()
+		if outboundPrivileges != nil {
+			return &sdk.GrantOwnershipOptions{
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:564]: it would be easier to have just one return in this method, so:
1. create options at the start
2. set outbound privileges inside if
3. return at the end

File: pkg/resources/grant_ownership.go [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+				Summary:  "Couldn't find OWNERSHIP privilege on target object",
+				Detail:   fmt.Sprintf("Id: %s", d.Id()),
+			},
+		}
+	}
+
+	return nil
+}
+
+func getOnObjectIdentifier(objectType sdk.ObjectType, objectName string) (sdk.ObjectIdentifier, error) {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:453]: this looks lika a general use validation method that could be inside the SDK (so each object has its dedicated type of identifier). Please leave a TODO with identifiers rework issue number and add reference to this method inside the issue's description, so that we don't forget about it.

File: pkg/resources/grant_ownership_test.go [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+		{
+			Name:       "validation - unsupported type",
+			ObjectType: sdk.ObjectTypeShare,
+			ObjectName: "some_share",
+			Error:      "object_type SHARE is not supported",
+		},
+		{
+			Name:       "validation - invalid account object identifier",
+			ObjectType: sdk.ObjectTypeDatabase,
+			ObjectName: "test_database.test_schema",
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_test.go:70]: We recently had the discussion about the validations for account object identifiers that concluded that it should be less strict for account identifiers, right? What were the arrangements there? Is it a topic that we will revisit with identifiers rework or...?

File: pkg/resources/grant_ownership.go [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+
+	if grantOn.Future != nil {
+		// TODO (SNOW-1182623): Still waiting for the response on the behavior/SQL syntax we should use here
+	} else {
+		accountRoleName, err := client.ContextFunctions.CurrentRole(ctx)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		err = client.Grants.GrantOwnership( // TODO: Should we always set outbound privileges to COPY in delete operation or set it to the config value?
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:354]: this is similar question like the one we had in copyGrants in other resources, right?

File: pkg/resources/grant_ownership.go [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+			ctx,
+			grantOn,
+			sdk.OwnershipGrantTo{
+				AccountRoleName: sdk.Pointer(sdk.NewAccountObjectIdentifier(accountRoleName)), // TODO: What if current role is database role (not a common but possible case)
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:358]: is it possible really? the use role is listed under account-level role management: https://docs.snowflake.com/en/sql-reference/sql/use-role. Are there any docs for use database role/ use role <database role>? If not, I think it deserves an internal question first.

File: pkg/resources/grant_ownership.go (hunk 7 above)
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:354]: how we know that this operation will succeed? What are the requirements for the tf role used for this to succeed? Can we explicitly state it at least in the docs (an ideally later maybe even validate this first)?

File: pkg/resources/grant_ownership.go [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+				ownershipFound = true
+			}
+		case ToDatabaseGrantOwnershipTargetRoleKind:
+			if grant.GranteeName.Name() == id.DatabaseRoleName.Name() {
+				ownershipFound = true
+			}
+		}
+	}
+
+	if !ownershipFound {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:440]: so should it be an error? in other (normal) resources, if we don't find the object we mark it as removed based on the error. So why is the no ownership an error. In the following example:
1. config with ownership
2. manual ownership transfer
3. terraform plan

the result would be an error. Should it be an error or should we try to set the ownership with tf?

File: pkg/resources/grant_ownership.go [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+	onObjectName := on["object_name"].(string)
+	onAll := on["all"].([]any)
+	onFuture := on["future"].([]any)
+
+	switch {
+	case len(onObjectType) > 0 && len(onObjectName) > 0:
+		objectType := sdk.ObjectType(strings.ToUpper(onObjectType))
+		objectName, err := getOnObjectIdentifier(objectType, onObjectName)
+		if err != nil {
+			return ownershipGrantOn, err
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership.go:531]: not `nil, err` to be more explicit? (even though underneath there is a nil)

File: pkg/resources/grant_ownership_test.go [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+			config := make(map[string]any)
+			if tt.AccountRole != nil {
+				config["account_role_name"] = *tt.AccountRole
+			}
+			if tt.DatabaseRole != nil {
+				config["database_role_name"] = *tt.DatabaseRole
+			}
+			d := schema.TestResourceDataRaw(t, grantOwnershipSchema, config)
+
+			defer func() {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_test.go:456]: parametrized tests are nice but if the setup becomes like this, I think it's more readable (and less error-prone) to setup to set of tests: one for valid and second for invalid configs. In this case for invalid ones we could just use assert/require.Panics(). Also, there is clear separation between db and account roles, so I think that splitting it into 3 different sets would be best, the ost readable option

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+
+	acc "github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/acceptance"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/hashicorp/terraform-plugin-testing/config"
+	"github.com/hashicorp/terraform-plugin-testing/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/tfversion"
+)
+
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:20]: please add a few negative tests with bad configs.

Also, the test if something happens outside of terraform (but I think such a test will be added in the followup PR with more sophisticated edge cases).

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/resources/grant_ownership.go [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/logging"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/provider"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+var validGrantOwnershipObjectTypes = []sdk.ObjectType{
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:19]: One question, after moving it to the `sdk`, should it be used as it is right now (resource-level validation), used to validate it on `sdk` level, or both `sdk` and resource level validation?

File: pkg/resources/grant_ownership.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:212]: I will try to set it to the NoOp function, but some time ago I had with difficulties with it, so nil != NoOp. Regardless, I'll add a note as to why there's no update.

File: pkg/resources/grant_ownership.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_ownership.go:19]: I think for now it's fine to have it on resource level only (we will think about duplicating it while extracting the SDK).

File: pkg/resources/grant_ownership.go (hunk 7 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:354]: Kind of, yeah. When someone defines this resource he/she will be thinking about `on create` effects setting COPY or REVOKE accordingly, but the question is: how do we behave when the resource is being removed? Should we transfer the ownership back with the same option as it was transferred in the Create or similarly in the case of the `view` resource it should be always internally set to COPY `on delete`?

File: pkg/resources/grant_ownership.go (hunk 8 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:358]: Hmm, I thought it's possible, but apparently it's not. I'll remove the comment.

File: templates/resources/grant_ownership.md.tmpl [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+where:
+- role_type - string - type of granted role (either ToAccountRole or ToDatabaseRole)
+- role_name - string - fully qualified identifier for either account role or database role (depending on the role_type)
+- outbound_privileges_behavior - string - behavior specified for existing roles (can be either COPY or REVOKE)
+- grant_type - enum
+- grant_data - data dependent on grant_type
+
+It has varying number of parts, depending on grant_type. All the possible types are:
+
+### OnObject
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply templates/resources/grant_ownership.md.tmpl:48]: Moved imports to imports.sh

File: pkg/sdk/object_types.go [hunk 15]
```diff
@@ -15,54 +15,59 @@ type Object struct {
...
+	ObjectTypeApplicationPackage   ObjectType = "APPLICATION PACKAGE"
+	ObjectTypeApplicationRole      ObjectType = "APPLICATION ROLE"
+	ObjectTypeStreamlit            ObjectType = "STREAMLIT"
+	ObjectTypeColumn               ObjectType = "COLUMN"
+	ObjectTypeIcebergTable         ObjectType = "ICEBERG TABLE"
+	ObjectTypeExternalVolume       ObjectType = "EXTERNAL VOLUME"
+	ObjectTypeNetworkRule          ObjectType = "NETWORK RULE"
+	ObjectTypePackagesPolicy       ObjectType = "PACKAGES POLICY"
+	ObjectTypeComputePool          ObjectType = "COMPUTE POOL"
+	ObjectTypeAggregationPolicy    ObjectType = "AGGREGATION POLICY"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/object_types.go:66]: SNOW-1230423 created to address this.

File: pkg/resources/grant_ownership.go (hunk 2 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:180]: Added link at the end.

File: pkg/resources/grant_ownership.go (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:379]: As discussed, every time where we cannot find the ownership on the target object

File: pkg/resources/grant_ownership.go (hunk 7 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:354]: Do you mean what would happen in cases where, e.g. the terraform role is not set to accountadmin?

File: pkg/resources/grant_ownership.go (hunk 9 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:440]: As discussed, we should set id to an empty string, and in the create operation we should try to forcefully take ownership and then grant it (or just grant it straight to the target role). Maybe to make it simpler I'll add it to the list of things I'll do in the next pr (should be added to the pr desc).

File: pkg/resources/grant_ownership_test.go (hunk 11 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_test.go:456]: Divided based on valid / invalid and role type

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 12 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:20]: Yeah, added to the list in the pr description to overall test outside of terraform interactions.

File: pkg/resources/grant_ownership.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership.go:212]: Noop cannot be set, because all of the fields are marked as ForceNew. I just left the comment why there's no Update.

File: pkg/resources/grant_ownership_test.go (hunk 6 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_test.go:70]: Yeah, that's my bad. This should be possible, so I'll change it and I'll add TODO to address this when we'll be working on identifiers.

File: pkg/resources/grant_ownership_test.go [hunk 16]
```diff
@@ -429,47 +455,33 @@ func TestGetOwnershipGrantTo(t *testing.T) {
...
 		})
 	}
 }
 
+func TestInvalidDatabaseRoleGetOwnershipGrantTo(t *testing.T) {
+	d := schema.TestResourceDataRaw(t, grantOwnershipSchema, map[string]any{
+		"database_role_name": "account_role_name",
+	})
+
+	defer func() {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_test.go:477]: could be done with assert.Panics as suggested, you can change in the follow-up PR.

**Review by sfc-gh-asawicki — DISMISSED** [review]

File: pkg/resources/grant_ownership_test.go (hunk 16 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_test.go]: Changed

**Review by sfc-gh-asawicki — DISMISSED** [review]

**Review by sfc-gh-asawicki — APPROVED** [review]
